core/src/apps/management/reboot_to_bootloader.py

```diff
@@ -7,20 +7,65 @@
 
 
 async def reboot_to_bootloader(msg: RebootToBootloader) -> NoReturn:
-    from trezor import io, loop, utils
+    from trezor import io, loop, utils, wire
+    from trezor.enums import BootCommand
     from trezor.messages import Success
     from trezor.ui.layouts import confirm_action
     from trezor.wire.context import get_context
 
-    await confirm_action(
-        "reboot",
-        "Go to bootloader",
-        "Do you want to restart Trezor in bootloader mode?",
-        verb="Restart",
-    )
-    ctx = get_context()
-    await ctx.write(Success(message="Rebooting"))
-    # make sure the outgoing USB buffer is flushed
-    await loop.wait(ctx.iface.iface_num() | io.POLL_WRITE)
-    utils.reboot_to_bootloader()
-    raise RuntimeError
+    if msg.boot_command is None or msg.boot_command == BootCommand.STOP_AND_WAIT:
+        await confirm_action(
+            "reboot",
+            "Go to bootloader",
+            "Do you want to restart Trezor in bootloader mode?",
+            verb="Restart",
+        )
+        ctx = get_context()
+        await ctx.write(Success(message="Rebooting"))
+        # make sure the outgoing USB buffer is flushed
+        await loop.wait(ctx.iface.iface_num() | io.POLL_WRITE)
+        utils.reboot_to_bootloader(BootCommand.STOP_AND_WAIT)
+        raise RuntimeError
+
+    elif (
+        msg.boot_command == BootCommand.INSTALL_UPGRADE
+        and msg.firmware_header is not None
+    ):
+        # check and parse received firmware header
+        hdr = utils.check_firmware_header(msg.firmware_header)
+        if hdr is None:
+            raise wire.ProcessError("Invalid firmware header.")
+        else:
+            # vendor must be the same
+            if hdr["vendor"] != utils.firmware_vendor():
+                raise wire.ProcessError("Different firmware vendor.")
+
+            current_version = (
+                int(utils.VERSION_MAJOR),
+                int(utils.VERSION_MINOR),
+                int(utils.VERSION_PATCH),
+            )
+
+            # firmware must be newer
+            if hdr["version"] <= current_version:
+                raise wire.ProcessError("Not a firmware upgrade.")
+
+            version_str = ".".join(map(str, hdr["version"]))
+
+            await confirm_action(
+                "reboot",
+                "FIRMWARE UPDATE",
+                f"Install firmware updated? Firmware version {version_str} by {hdr['vendor']}",
+                verb="INSTALL",
+            )
+
+            ctx = get_context()
+            await ctx.write(Success(message="Rebooting"))
+            # make sure the outgoing USB buffer is flushed
+            await loop.wait(ctx.iface.iface_num() | io.POLL_WRITE)
+            # reboot to the bootloader, pass the firmware header hash
+            utils.reboot_to_bootloader(BootCommand.INSTALL_UPGRADE, hdr["hash"])
+            raise RuntimeError
```

Comment:
this chunk of code should be shared between the two branches.
(so probably each branch sets `boot_command = X` and `boot_args = Y`, and then the common tail invokes `reboot_to_bootloader(boot_command, boot_args)`?)

Reply:
changed ProcessError to DataError, see 1bdf75b75a876a748fe92c1d7d01dfd62301efd0

Reply:
@mmilata fixed the rest, see f62afd3947755a0645b92be72ca3208b175cdf2e